File: static/js/store.js
/* Persistence.
 *
 * Everything the app knows lives in this browser. An early version of the
 * server kept pilot channels in memory only: a restart silently reset them to
 * defaults, and the next config write pushed those defaults onto the timer,
 * overwriting the real race frequencies. Anything that can be written to the
 * hardware is saved here first.
 */
'use strict';

const NS = 'wt.';

export function load(key, fallback) {
  try {
    const raw = localStorage.getItem(NS + key);
    return raw == null ? fallback : (JSON.parse(raw) ?? fallback);
  } catch (e) { return fallback; }
}

export function save(key, value) {
  try { localStorage.setItem(NS + key, JSON.stringify(value)); } catch (e) {}
  return value;
}

export function clear(key) {
  try { localStorage.removeItem(NS + key); } catch (e) {}
}

export const DEFAULT_SETTINGS = {
  mode: 'practice',
  targetLaps: 5,
  targetSeconds: 120,
  consecN: 3,
  minLap: 3.0,
  holeshot: false,
  countdown: 5,
  timerMinLapMs: 3000,
  preset: 'normal',
};

export const DEFAULT_PREFS = {
  voiceOn: true,
  voiceName: '',
  rate: 1.1,
  announce: 'full',      // full: "Lap 3, 24.7" | time: "24.7" | off
  theme: 'dark',
  keepAwake: true,
};

export function settings() { return { ...DEFAULT_SETTINGS, ...load('settings', {}) }; }
export function prefs() { return { ...DEFAULT_PREFS, ...load('prefs', {}) }; }

/** Save a finished session. A race that ended, had its last lap undone and
 *  ended again is one race, not two: an entry with the same runId is replaced. */
export function appendHistory(entry, cap = 200) {
  const h = load('history', []);
  entry.savedAt = Date.now() / 1000;
  const i = entry.runId ? h.findIndex(e => e.runId === entry.runId) : -1;
  if (i >= 0) h[i] = entry; else h.push(entry);
  return save('history', h.slice(-cap));
}

/** CSV of every saved session, the same shape the Python build exported. */
export function historyCsv() {
  const rows = [['session', 'when', 'mode', 'pos', 'pilot', 'channel',
                 'laps', 'best', 'best_consec', 'total', 'lap_times']];
  for (const r of load('history', [])) {
    const when = new Date((r.at || 0) * 1000).toISOString().slice(0, 16).replace('T', ' ');
    for (const e of r.results || []) {
      rows.push([r.name, when, r.mode, e.pos, e.name, e.channel, e.laps,
                 e.best ?? '', e.consec ?? '', e.total ?? '',
                 (e.lapTimes || []).map(x => x.toFixed(2)).join(' ')]);
    }
  }
  return rows.map(r => r.map(cell => {
    const s = String(cell ?? '');
    return /[",\n]/.test(s) ? '"' + s.replace(/"/g, '""') + '"' : s;
  }).join(',')).join('\n');
}

File: store.py
"""Persistence. Everything the app knows survives a restart.

This exists because an early version kept pilot channels in memory only: a
server restart silently reset them to defaults, and the next config write
pushed those defaults onto the timer, overwriting the real race frequencies.
Anything that can be written to the hardware is persisted here first.
"""
import json, os, time, threading

DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")
LOCK = threading.Lock()

DEFAULTS = {
    "pilots": {},        # slot -> {name, channel, enabled, colour}
    "roster": [],        # [{name, channel, colour}] reusable pilot list
    "rf": {},            # slot -> {gain, threshold}
    "settings": {
        "mode": "laps", "targetLaps": 5, "targetSeconds": 120,
        "minLap": 3.0, "holeshot": False, "countdown": 5,
        "bestConsecutive": 3, "sigThreshold": 500.0, "autoDetect": False,
        "voice": True, "timerMinLapMs": 3000,
    },
    "history": [],       # completed races
}


def _path(name):
    os.makedirs(DIR, exist_ok=True)
    return os.path.join(DIR, name + ".json")


def load(name):
    try:
        with open(_path(name)) as f:
            return json.load(f)
    except Exception:
        d = DEFAULTS.get(name)
        return json.loads(json.dumps(d)) if d is not None else None


def save(name, obj):
    with LOCK:
        p = _path(name)
        tmp = p + ".tmp"
        with open(tmp, "w") as f:
            json.dump(obj, f, indent=2)
        os.replace(tmp, p)      # atomic: never leave a half-written config
    return obj


def append_history(entry, cap=200):
    h = load("history") or []
    entry["savedAt"] = time.time()
    h.append(entry)
    return save("history", h[-cap:])
